core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -292,34 +309,46 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    */
   private class Watchers(val key: Any) {
 
-    private[this] val operations = new LinkedList[T]()
+    private[this] val operations = new ConcurrentLinkedQueue[T]()
+    private[this] val approximateWatched = new AtomicInteger(0)
```

Comment:
This is to avoid the linear `size()` of `ConcurrentLinkedQueue`. It is only used for purgatory metrics, so if we are not too concerned about the performance, we could drop it (this would also allow us to cleanup some of the new purged logic).